### DISCUSSION TITLE
Using fsspec to generalize FileSystemProvider

### DISCUSSION THREAD

**Discussion by ymoisan** [geopython/pygeoapi#1189] [Ideas]
Would it make sense to use [fsspec](https://filesystem-spec.readthedocs.io/en/latest/) to make the [FileSystemProvider](https://github.com/geopython/pygeoapi/blob/master/pygeoapi/provider/filesystem.py) able to retrieve files also from urls and specific cloud provider object storage ?

``` python
# Open local file
fsspec_local = fsspec.open('.../pygeoapi/tests/data/poi_portugal.gpkg')
fsspec_local
<OpenFile '.../pygeoapi/tests/data/poi_portugal.gpkg'>
# Open a file on Azure blob storage via url; can be done with `az://` instead of `https`
fsspec_az_http = fsspec.open('https://....blob.core.windows.net/.../somefile.xml')
fsspec_az_http
<OpenFile 'https://....blob.core.windows.net/.../somefile.xml'>
# Open a file on S3;  can be done with `s3://` instead of `https`
url = 'https://mur-sst.s3.us-west-2.amazonaws.com/zarr-v1' # Amazon Public Data
fsspec_s3 = fsspec.open(url)
fsspec_s3
<OpenFile 'https://mur-sst.s3.us-west-2.amazonaws.com/zarr-v1'>
```

**Comment by ymoisan** [comment]
One significant advantage I see is for read access on public object storage.  No more credentials to shove around.  Just urls.

**Reply by francbartoli** [reply]
Thanks @ymoisan, it is on my todo list. Can you file a ticket for this, I will flag it as improvement and let you know. Many thanks for your interest in pygeoapi
